lib/probes/koa-router.js

```diff
@@ -2,67 +2,166 @@
 
 const shimmer = require('ximmer')
 const methods = require('methods')
+const semver = require('semver')
+
 const utility = require('../utility')
 const ao = require('../')
+const requirePatch = require('../require-patch')
 const Span = ao.Span
 const conf = ao.probes['koa-router']
 
-module.exports = function (router) {
-  return app => patchApp(app, router)
+const {version} = requirePatch.relativeRequire('koa-router/package.json')
+
+const logMissing = ao.makeLogMissing('koa-router')
+
+module.exports = function (_router) {
+
+  if (semver.gte(version, '6.0.0')) {
+    // allow the caller to invoke this with or without "new" by not using
+    // an arrow function.
+    return function () {
+      const router = new _router(...arguments)
+
+      methods.concat('del', 'all').forEach(method => {
+        patchAsyncMethod(router, method)
+      })
+      return router
+    }
+  }
+
+  const v5 = semver.gte(version, '5.0.0')
+
+  // version 5 and below
+  return function (app) {
+    const router = _router(app)
+    // For 5.x+, the methods are on the router, not the app, so accomodate both.
+    const target = v5 ? router : app
+    if (!target) {
+      logMissing('target to patch')
+      return router
+    }
+
+    methods.concat('del').forEach(method => {
+      patchGeneratorFunction(target, method)
+    })
+    return router
+  }
 }
 
-function patchApp (app, router) {
-  const handler = router(app)
+//
+// v6+ sequence: patchAsyncMethod, patchAsyncHandler
+//
+function patchAsyncMethod (target, method) {
+  if (typeof target[method] !== 'function') {
+    logMissing(method + '()')
+    return
+  }
+
+  shimmer.wrap(target, method, fn => {
+    return function (name, url, route) {
+      if (arguments.length === 3) {
+        return fn.call(this, name, url, patchAsyncHandler(
+          method + ' ' + url,
+          utility.fnName(route),
+          route
+        ))
+      }
 
-  methods.concat('del').forEach(method => {
-    // For 5.x+, the methods are on a router, not patched onto the app
-    //Check if app exists, only use handler otherwise
-    patchMethod(
-      (app ? [handler, app] : [handler]).filter(v => typeof v[method] === 'function')[0],
-      method
-    )
-  })
+      if (arguments.length !== 2) {
+        logMissing('correct number of arguments, found %d', arguments.length)
```

Comment:
just a stylistic suggestion for a more sequential ordering :
if (arguments.length === 3) { .... }
if (arguments.length === 2) { .... }
logMissing

Reply:
studied the tests as well
lgtm :)